adapters/Vungle/VungleAdapter/GADMAdapterVungleRouter.m

```diff
@@ -126,67 +126,87 @@ - (BOOL)addDelegate:(nonnull id<GADMAdapterVungleDelegate>)delegate {
+          if ([vungleBannerRequest.placementID isEqualToString:delegate.bannerRequest.placementID]) {
             return NO;
-          } else {
-            NSOrderedSet *set = [_bannerDelegates objectForKey:bannerRequest];
-            for(id<GADMAdapterVungleDelegate> delegate in set) {
-              if (delegate.bannerState == BannerRouterDelegateStateWillPlay || delegate.bannerState == BannerRouterDelegateStatePlaying) {
-                return NO;
-              }
+          } else if (vungleBannerRequest.placementID.length > 0 && ![vungleBannerRequest.placementID isEqualToString:delegate.bannerRequest.placementID]) {
+            // There is already a banner delegate with different placementID in _bannerDelegates.
+            if (!_bannerRequest) {
+              _bannerRequest = [vungleBannerRequest copy];
             }
+            return NO;
           }
         }
 
-        NSMutableOrderedSet *set = [[NSMutableOrderedSet alloc] initWithObject:delegate];
-        GADMAdapterVungleMapTableSetObjectForKey(_bannerDelegates, delegate.bannerRequest, set);
-        _bannerPlacementID = [delegate.desiredPlacement copy];
-      } else if (delegate && [_bannerDelegates objectForKey:delegate.bannerRequest]) {
-        NSMutableOrderedSet *set = [[_bannerDelegates objectForKey:delegate.bannerRequest] mutableCopy];
-        if (![set containsObject:delegate]) {
-          GADMAdapterVungleMapTableRemoveObjectForKey(_bannerDelegates, delegate.bannerRequest);
-          [set addObject:delegate];
-          GADMAdapterVungleMapTableSetObjectForKey(_bannerDelegates, delegate.bannerRequest, set);
-          _bannerPlacementID = [delegate.desiredPlacement copy];
-        }
+        GADMAdapterVungleMapTableSetObjectForKey(_bannerDelegates, delegate.bannerRequest, delegate);
+        _bannerRequest = [delegate.bannerRequest copy];
+      } else {
+        id<GADMAdapterVungleDelegate> bannerDelegate = [_bannerDelegates objectForKey:delegate.bannerRequest];
+
+        /* The isRequestingBannerAdForRefresh flag is used for an edge case, when the old Banner
+         * delegate is removed from _bannerDelegates and there is a refresh Banner delegate doesn't
+         * construct Banner view successfully and add to _bannerDelegates yet. Adapter cannot set
+         * _bannerRequest to nil to prevent requesting another Banner ad with different bannerRequest.
+         */
+        bannerDelegate.isRequestingBannerAdForRefresh = YES;
+        delegate.isRefreshedForBannerAd = YES;
       }
     }
   }
   return YES;
 }
 
+- (void)replaceOldBannerDelegate:(nonnull id<GADMAdapterVungleDelegate>)delegate {
+  if (!delegate) {
+    return;
+  }
+
+  if ([delegate respondsToSelector:@selector(isBannerAd)] && [delegate isBannerAd]) {
+    @synchronized(_bannerDelegates) {
+      // We only support displaying one Vungle Banner Ad at the same time currently
+      if (_bannerRequest != nil && ![_bannerRequest isEqualToBannerRequest:delegate.bannerRequest]) {
```

Comment:
At this point where this method is called, we've already allowed a new ad to load. If we quit out here and don't let the delegate get overridden, then we'll be sending events to the old delegate instance even though this is for the ad refresh. I think it's too late to make this check here.

Reply:
@ericleich Hi Eric, thanks for the good suggestion,  I agree with you. This is just a safety check, just don't want to add a delegate with different bannerRequest. For a refresh banner delegate, that rarely happens, because the _bannerRequest should be always equal to the bannerRequest of a refresh banner delegate.